snapcraft/internal/remote_build/_worktree.py

```diff
@@ -14,100 +14,263 @@
-                    logger.debug("Remove file: {}".format(dest))
-                    self._unlink(dest)
-
-            for name in dirnames:
-                dest = os.path.join(dirpath, name)
-                src = os.path.join(self._srcpath(dirpath), name)
-                if not os.path.exists(src) and os.path.exists(dest):
-                    logger.debug("Remove directory: {}".format(dest))
-                    os.rmdir(dest)
-
-        # finally commit our changes, if any
-        self._repo.commit()
-
-    def add_remote(self, provider: str, user: str, build_id: str) -> str:
-        """Add a remote repository."""
-        return self._repo.add_remote(provider, user, build_id)
-
-    def push(self) -> None:
-        """Send our data to a remote repository."""
-        self._repo.push()
-
-    def _ignored_files(self, source: str, contents: List[str]) -> List[str]:
-        ignored = set()
-        for pattern in self._ignore:
-            ignored.update(set(fnmatch.filter(contents, pattern)))
-        return list(ignored)
-
-    def _link_and_add(self, source: str, destination: str) -> None:
-        link_or_copy(source, destination)
-        self._repo.add(os.path.relpath(destination, self._destroot))
-        logger.debug("Add file {}".format(destination))
-
-    def _unlink(self, path: str) -> None:
-        self._repo.remove(os.path.relpath(path, self._destroot))
-
-    def _srcpath(self, path: str) -> str:
-        rel = os.path.relpath(path, self._destroot)
-        return os.path.join(self._srcroot, rel)
+        return self._git_repo
+
+    def _process_part_sources(self, part_name: str, part_yaml: OrderedDict):
+        """Add part sources to tree.  Return source-modified yaml for part."""
+        logger.debug("processing {}: {}".format(part_name, part_yaml))
+
+        if "source" not in part_yaml:
+            logger.debug("no source specified, skipping...")
+            return part_yaml
+
+        source = part_yaml["source"]
+        source_modified = False
+        if isinstance(source, list):
+            # Grammar case.
+            part_yaml["source"] = []
+            for s in source:
+                sinfo = OrderedDict()  # type: OrderedDict
+                for selector, selector_source in s.items():
+                    sinfo[selector] = self._pull_source(
+                        part_name, selector_source, selector=selector
+                    )
+                    if sinfo[selector] != selector_source:
+                        source_modified = True
+
+                part_yaml["source"].append(sinfo)
+        else:
+            part_yaml["source"] = self._pull_source(part_name, source)
+            if part_yaml["source"] != source:
+                source_modified = True
+
+        if source_modified:
+            # It's now an archive.
+            part_yaml["source-type"] = "tar"
+
+            # Strip source attributes no longer relevant.
+            for key in [
```

Comment:
we should add a TODO to make this more structured and pick these up from the actual source handler or schema considering that we will be finding ourselves in a potential place where we add a new key and forget to pop it here.

Reply:
I changed it to instead just pop all keys that start with "source-", and then set the source attributes we dictate here (i.e. source-type).